libtenzir/builtins/operators/grok.cpp

```diff
@@ -0,0 +1,518 @@
+    }
+    result_pattern.append(previous_match_end, str_re2.end());
+    resolved_pattern.emplace(std::move(result_pattern));
+    TENZIR_ASSERT(resolved_pattern);
+    TENZIR_ASSERT(not resolved_pattern->empty());
+  }
+}
+
+void pattern_store::parse_line(std::string_view line) {
+  if (line.empty())
+    return;
+  if (line.starts_with('#'))
+    return;
+  auto parts = detail::split(line, " ", 1);
+  TENZIR_ASSERT(not patterns.contains(std::string{parts[0]}));
+  patterns.emplace(std::string{parts[0]}, std::string{parts[1]});
+}
+
+auto& get_builtin_pattern_store() {
+  // Using CoW to avoid copying the builtin patterns unless we need to
+  static auto store = caf::make_copy_on_write<pattern_store>(
+    std::span{builtin_patterns_strings});
+  return store;
+}
+
+class grok_parser final : public plugin_parser {
+public:
+  grok_parser() = default;
+
+  explicit grok_parser(parser_interface& p)
+    : patterns_(get_builtin_pattern_store()) {
+    auto parser
+      = argument_parser{"grok", "https://docs.tenzir.com/next/operators/"
+                                "transformations/grok"};
+    parser.add(input_pattern_.raw_pattern, "<input_pattern>");
+    std::optional<std::string> pattern_definitions{};
+    parser.add("--pattern-definitions", pattern_definitions, "<patterns>");
+    parser.add("--indexed-captures", indexed_captures_);
+    parser.add("--include-unnamed", include_unnamed_);
+    parser.add("--include-nulls", include_nulls_);
+    parser.parse(p);
+    if (pattern_definitions)
+      patterns_.unshared().add(*pattern_definitions);
+    TENZIR_ASSERT_EXPENSIVE(std::ranges::all_of(
+      patterns_->patterns | std::views::values, [](const auto& p) -> bool {
+        return p.resolved_pattern.has_value();
+      }));
+    input_pattern_.resolve(*patterns_, false);
+  }
+
+  auto name() const -> std::string override {
+    return "grok";
+  }
+
+  auto
+  instantiate(generator<chunk_ptr> input, operator_control_plane& ctrl) const
+    -> std::optional<generator<table_slice>> override {
+    (void)input;
+    diagnostic::error("`{}` cannot be used here", name())
+      .emit(ctrl.diagnostics());
+    return {};
+  }
+
+  auto parse_strings(std::shared_ptr<arrow::StringArray> input,
+                     operator_control_plane& ctrl) const
+    -> std::vector<std::pair<type, std::shared_ptr<arrow::Array>>> override {
+    auto builder = series_builder{type{record_type{}}};
+    for (auto&& string : values(string_type{}, *input)) {
+      if (not string) {
+        builder.null();
+        continue;
+      }
+      boost::cmatch matches{};
+      if (not boost::regex_match(string->begin(), string->end(), matches,
+                                 *input_pattern_.resolved_pattern)) {
+        diagnostic::warning("pattern could not be matched")
+          .hint("input: `{}`", string)
+          .hint("pattern: `{}`", input_pattern_.resolved_pattern->str())
+          .emit(ctrl.diagnostics());
```

Comment:
This could be very noisy (in particular if you change the `return` to `continue`, which it arguably should be), but I guess that's fine for now, given that we currently only focus on the happy path of `parse`. You could perhaps even make this an `error`.

Reply:
Yeah, sure, let's make it an error to avoid everything blowing up in `parse`.

Reply:
If you add a `null` it should not blow up.